interface/forms/annotate_diagram/mapdiagram/js/mapdiagram.js

```diff
@@ -0,0 +1,333 @@
+        var elem = document.getElementById(elval);
+        var style = window.getComputedStyle(elem, null).getPropertyValue('font-size');
+        var fontsz = parseInt(style);
+        var cpos = new Object();
+        if( fontsz < 12) fontsz = 12;
+        cpos[0] = xpos - (fontsz / 2)-1; 
+        cpos[1] = ypos - ((fontsz+(fontsz / 3))/2);
+        cpos[1] =  Math.round(cpos[1]);
+        if(tpos == '░'){
+            cpos[0] = cpos[0] + 2; 
+            cpos[1] = cpos[1] + 4;
+        }
+        
+        return cpos;
+    };
+    var fn_setMode = function () {
+        if (gf){
+            gf = false; isInput = false; prx=0; pry=0;
+            txtmode = false;
+            $(".dytxt").val("");
+            $('.dytxt').hide();
+            //$("#mode").attr("style", "display:none;");
+            $("#btn_mode").attr("style", "color:black;font-weight:normal;");
+            $("#btn_mode").text('Legend Mode');
+            $(".mode-inline").text('Mode Label');
+            $('#container').removeClass('symcursor');
+            mode_styleEl.visibility = "hidden";
+        }
+        else{
+            gf = true;
+            //$('#rtxt').attr('checked', 'checked');
+            //fn_setMarkval();
+            $("#btn_mode").attr("style", "color:red;");
+            //$("#mode").attr("style", "display:inline-block;");
+            $("#btn_mode").text("Label Mode");
+        }
+        
+    };
+    var fn_setMarkval = function () {
+    	$('.dytxt').hide();
+    	mode_obj = $('input[name=modegrp]:checked').val();
+    	
+         if( mode_obj == "txtmode" ) {
+        	$('#container').removeClass('symcursor');
+            txtmode = true;
+            mode_obj = "";
+            }
+        else{
+        	$('#container').addClass('symcursor');
+            txtmode = false; isInput = false;
+            prx=0; pry=0; 
+            }
+        
+        if(txtmode) $(".mode-inline").text('Mode Legend : Text');
+    	else $(".mode-inline").text('Mode Legend : '+ mode_obj);
+        if(!gf){ // in label mode
+    		fn_setMode();
+    	}
+    };
+    /* main */
+    var dropdownOptions = args.dropdownOptions;
+    var options = dropdownOptions.options;
+    var optionsLabel = dropdownOptions.label;
+    var container = args.container;
+    var data = args.data;
+    var hideNav = args.hideNav;
+
+    container.mousemove( function (){ f = true; } );
+    $('.dytxt').blur( function(){ isInput = true; } );
+ 
+    if (!hideNav) {
+        container.click(function (e) {
+            if (!f) return;
+               
+            var x = e.pageX - this.offsetLeft - 5;
+            var y = e.pageY - this.offsetTop - 5;
+            var dialog = $(".dialog-form").clone();
+            dialog.find(".label").val(counter + 1);
+            var hasOptions = typeof (options) != "undefined";
+            if (hasOptions) {
+                dialog.find("label[for='options']").text(typeof (optionsLabel) != "undefined" ? optionsLabel : "Select one");
```

Comment:
Generally best to avoid strings in js, if possible, since it is tougher to translate. Albeit it can be done.